### ISSUE TITLE
Possibly broken AIS tests and implementation

### ISSUE THREAD

**Issue by artsobolev** [tensorflow/probability#267]
I was trying to get `sample_annealed_importance_chain` to work on my problem and looked at how the interface is supposed to be used in test, and found out a weird line of code that does not seem to agree with theory, nor agree with implied interface.

In [sample_annealed_importance_test.py:109](https://github.com/tensorflow/probability/blob/de7907e18d94a1d3dff0c0294e6a1ddd20402656/tensorflow_probability/python/mcmc/sample_annealed_importance_test.py#L109) test essentially calculates `tf.log(1/num_steps * tf.reduce_sum(tf.exp(ais_weights)))`, which looks like averaging in exp-space and then going to log-space back, but the problem is that `ais_weights` has shape of `init`, and it bears no relation to `num_steps` whatsoever. I guess it was actually meant to run several independent chains, but then one should normalize by number of elements in `init` (which [seems](https://github.com/tensorflow/probability/blob/de7907e18d94a1d3dff0c0294e6a1ddd20402656/tensorflow_probability/python/mcmc/sample_annealed_importance_test.py#L147) to have just 60 elements, which is off from log(200) by ~1.2 nats.

This and my own failures at getting AIS to estimate log-likelihood (following [Wu et al.](https://arxiv.org/abs/1611.04273)) have led me to suspicion that the current implementation is bugged, and the bugs in the implementation and tests just cancel each other out. In particular [this line](https://github.com/tensorflow/probability/blob/0f5dae99f1161c1b4050b6d503b2607129fba8a8/tensorflow_probability/python/mcmc/sample_annealed_importance.py#L222) seems troubling: one just reevaluates the same log-densities over and over at new samples, whereas AIS requires one to use different log-probabilities at different steps (see formulas 3-5 of [the paper](https://arxiv.org/pdf/physics/9803008.pdf)). I also don't think the averaging over `num_steps` is necessary, as for larger number of steps consecutive log-densities should be closer, contributing less to the final estimate.

**Comment by brianwa84** [comment]
@SiegeLordEx can you take a look?

**Comment by artsobolev** [comment]
Ok, now I think the implementation is correct, but the test is still puzzling.

**Comment by SiegeLordEx** [comment]
Thanks for investigating, @artsobolev. I am pretty confident that the implementation is correct it just looks weird because it's in logspace, but as you say that test line is nonsense. It should be `log(num_chains)` as you say. The reason why the test doesn't fail is because it's testing the raw AIS weights. The `log_estimated_normalizer` is only logged, and not checked at all.

That line should be fixed and then either removed, or perhaps we can write the test in terms of checking for its value.
